UI/src/main/java/adList/Ad.java

```diff
@@ -0,0 +1,157 @@
+package adList;
+
+import com.google.gson.Gson;
+
+import java.util.*;
+
+public class Ad {
+    private final String id;
+    private String description;
+    private final Date createdAt;
+    private String link;
+    private String vendor;
+    private String photoLink;
+    private List<String> hashTags;
+    private String discount;
+    private Date validUntil;
+    private Double rating;
+    private List<String> reviews;
+
+    public Ad(String id, String description, Date createdAt, String link, String vendor, List<String> hashTags, String discount, Date validUntil) {
+        this.id = id;
+        this.description = description;
+        this.createdAt = createdAt != null ? (Date) createdAt.clone() : null;
+        this.link = link;
+        this.vendor = vendor;
+        this.hashTags = hashTags != null ? new ArrayList<>(hashTags) : null;
+        this.discount = discount;
+        this.setValidUntil(validUntil);
+    }
+    public Ad(Ad ad){
+        if(ad != null) {
```

Comment:
Putting a lot of code in `{ ... }` deteriorates reading. Extra indents, extra braces... Therefore, it's usually recommended to use the "early fail - early return" technique. That is, instead of
```
if (id != null) {
    ...
    ...
    ...
    ...
} 
```
write
```
if (id == null) {
   return;
}
// here comes everything else. We are done with if-ing.
```